pkg/plugins/policies/meshratelimit/api/v1alpha1/validator.go

```diff
@@ -0,0 +1,115 @@
+package v1alpha1
+
+import (
+	"time"
+
+	common_api "github.com/kumahq/kuma/api/common/v1alpha1"
+	"github.com/kumahq/kuma/pkg/core/validators"
+	matcher_validators "github.com/kumahq/kuma/pkg/plugins/policies/matchers/validators"
+)
+
+func (r *MeshRateLimitResource) validate() error {
+	var verr validators.ValidationError
+	path := validators.RootedAt("spec")
+	if len(r.Spec.From) == 0 {
+		verr.AddViolationAt(path.Field("from"), "needs at least one item")
+	}
+	verr.AddErrorAt(path.Field("targetRef"), validateTop(r.Spec.TargetRef, hasTcpConfiguration(r.Spec.From)))
+	verr.AddErrorAt(path, validateFrom(r.Spec.From))
+	return verr.OrNil()
+}
+
+func validateTop(targetRef common_api.TargetRef, hasTcpConfig bool) validators.ValidationError {
+	supportedKinds := []common_api.TargetRefKind{
+		common_api.Mesh,
+		common_api.MeshSubset,
+		common_api.MeshService,
+		common_api.MeshServiceSubset,
+	}
+	if !hasTcpConfig {
+		supportedKinds = append(supportedKinds, common_api.MeshGatewayRoute)
+	}
+	targetRefErr := matcher_validators.ValidateTargetRef(targetRef, &matcher_validators.ValidateTargetRefOpts{
+		SupportedKinds: supportedKinds,
+	})
+	return targetRefErr
+}
+
+func validateFrom(from []From) validators.ValidationError {
+	var verr validators.ValidationError
+	for idx, fromItem := range from {
+		path := validators.RootedAt("from").Index(idx)
+		defaultField := path.Field("default")
+		verr.Add(validateDefault(defaultField, fromItem.Default))
+		hasTcpConfig := isTcp(fromItem.Default)
+		supportedKinds := []common_api.TargetRefKind{
+			common_api.Mesh,
+			common_api.MeshSubset,
+			common_api.MeshService,
+			common_api.MeshServiceSubset,
+		}
+		if hasTcpConfig {
+			supportedKinds = []common_api.TargetRefKind{
+				common_api.Mesh,
+			}
+		}
+		verr.AddErrorAt(path.Field("targetRef"), matcher_validators.ValidateTargetRef(fromItem.GetTargetRef(), &matcher_validators.ValidateTargetRefOpts{
+			SupportedKinds: supportedKinds,
+		}))
+	}
+	return verr
+}
+
+func validateDefault(path validators.PathBuilder, conf Conf) validators.ValidationError {
+	var verr validators.ValidationError
+	local := conf.Local
+	if local == nil {
```

Comment:
In practice you could make `local` non pointer. I think it'd make this a little simpler